buckets/local/bucket.go

```diff
@@ -243,29 +294,88 @@ func (b *Bucket) HashFile(pth string) (cid.Cid, error) {
 	if err != nil {
 		return cid.Undef, err
 	}
-	n, err := addFile(b.tmp, b.layout, prefix, r)
+	n, err := addFile(dagutils.NewMemoryDagService(), b.layout, prefix, r)
 	if err != nil {
 		return cid.Undef, err
 	}
 	return n.Cid(), nil
 }
 
-// Diff returns a list of changes that are present in pth compared to the current archive.
+// Diff returns a list of changes that are present in path compared to the bucket.
 func (b *Bucket) Diff(ctx context.Context, pth string) (diff []*dagutils.Change, err error) {
+	tmp := dagutils.NewMemoryDagService()
 	var an ipld.Node
-	if b.root.Defined() {
-		an, err = b.tmp.Get(ctx, b.root)
+	if b.local.Defined() {
+		an, err = b.dag.Get(ctx, b.local)
 		if err != nil {
 			return
 		}
 	} else {
 		an = unixfs.EmptyDirNode()
 	}
-	bn, err := b.recursiveAddPath(ctx, pth)
+	if err = copyLinks(ctx, an, b.dag, tmp); err != nil {
+		return
+	}
+	if err = tmp.Add(ctx, an); err != nil {
+		return
+	}
+	bn, err := b.recursiveAddPath(ctx, pth, tmp)
 	if err != nil {
 		return
 	}
-	return dagutils.Diff(ctx, b.tmp, an, bn)
+	return dagutils.Diff(ctx, tmp, an, bn)
+}
+
+// pathMap hold details about a local path map.
+type pathMap struct {
+	Remote    string
+	LocalCid  cid.Cid
+	RemoteCid cid.Cid
+}
+
+// AddPathMap adds a local path map to the store.
+func (b *Bucket) AddPathMap(lp, rp string, lc, rc cid.Cid) error {
```

Comment:
I haven't used this yet... ran out of time on this PR. This is related to the following problem:
In a private bucket, the remote is tracking encrypted cids while the local uses cids from the plaintext files. This works fine now, but we can't selectively _pull_ files, since the remote cids won't ever match the local ones. That were this comes in, allowing a local lookup of a remote cid -> local cid mapping. 

If I have time later, I'll try and get this in, otherwise, can wait until next PR next week.

Reply:
I just snuck this in... took a little more thought than I expected. Actually, the path maps were also needed to handle removing blocks from the flatfs repo.